codegen/winterfell/src/air/graph.rs

```diff
@@ -65,29 +67,62 @@ impl Codegen for NodeIndex {
+            Operation::Mul(_, _) => binary_op_to_string(ir, self, elem_type, trace_segment),
+            Operation::Exp(l_idx, r_idx) => {
+                let lhs = l_idx.to_string(ir, elem_type, trace_segment);
+                let lhs = if is_leaf(l_idx, ir) {
+                    lhs
+                } else {
+                    format!("({lhs})")
+                };
+                match r_idx {
+                    0 => match elem_type {
+                        // x^0 = 1
+                        ElemType::Base => "Felt::ONE".to_string(),
+                        ElemType::Ext => "E::ONE".to_string(),
+                    },
+                    1 => lhs, // x^1 = x
+                    _ => match elem_type {
+                        ElemType::Base => format!("{lhs}.exp(Felt::new({r_idx}))"),
+                        ElemType::Ext => {
+                            format!("{lhs}.exp(E::PositiveInteger::from({r_idx}_u64))")
+                        }
+                    },
+                }
+            }
+        }
+    }
+}
+
+impl Codegen for Value {
+    fn to_string(&self, ir: &AirIR, elem_type: ElemType, trace_segment: u8) -> String {
+        match self {
+            Value::Constant(ConstantValue::Inline(0)) => match elem_type {
                 ElemType::Base => "Felt::ZERO".to_string(),
                 ElemType::Ext => "E::ZERO".to_string(),
             },
-            Operation::Constant(ConstantValue::Inline(1)) => match elem_type {
+            Value::Constant(ConstantValue::Inline(1)) => match elem_type {
                 ElemType::Base => "Felt::ONE".to_string(),
                 ElemType::Ext => "E::ONE".to_string(),
             },
-            Operation::Constant(ConstantValue::Inline(value)) => match elem_type {
+            Value::Constant(ConstantValue::Inline(value)) => match elem_type {
                 ElemType::Base => format!("Felt::new({value})"),
                 ElemType::Ext => format!("E::from({value}_u64)"),
             },
-            Operation::Constant(ConstantValue::Scalar(ident)) => match elem_type {
+            Value::Constant(ConstantValue::Scalar(ident)) => match elem_type {
                 ElemType::Base => ident.to_string(),
                 ElemType::Ext => format!("E::from({ident})"),
             },
-            Operation::Constant(ConstantValue::Vector(vector_access)) => match elem_type {
+            Value::Constant(ConstantValue::Vector(vector_access)) => match elem_type {
                 ElemType::Base => format!("{}[{}]", vector_access.name(), vector_access.idx()),
                 ElemType::Ext => {
                     format!("E::from({}[{}])", vector_access.name(), vector_access.idx())
                 }
             },
-            Operation::Constant(ConstantValue::Matrix(matrix_access)) => match elem_type {
+            Value::Constant(ConstantValue::Matrix(matrix_access)) => match elem_type {
```

Comment:
Yeah, I agree that it would. I did this work before the constant variants were split apart, so this is just the update that was made during the rebase. I'd prefer to address this separately in a future PR rather than revisiting this refactor now.

Reply:
I added a TODO